warble_server.cpp

```diff
@@ -193,9 +203,40 @@ std::optional<std::string> WarbleServer::Profile(
 // return the latest warble with given hashtag
 std::optional<std::string> WarbleServer::Stream(
     const google::protobuf::Any& payload) {
-  // TODO: Get latest warble with tag from kvstore
+  StreamRequest request;
+  payload.UnpackTo(&request);
+  if (!ValidateUser(request.username())) {
+    LOG(ERROR) << "the user " << request.username() << "is not registered";
+    return "the user is not registered";
+  }
+  Warble* warble = new Warble();
+  bool valid_warble = (*warble).ParseFromString(
+      kvstore_.Get(std::vector<std::string>(1, request.tag() + "_tag"))[0]);
+
+  // did not find warbles with this hashtag
+  if (!valid_warble) {
+    delete warble;
+    return std::nullopt;
+  }
+  // found a warble but it isn't the latest so don't stream
+  std::vector<std::string> time_history =
+      kvstore_.Get(std::vector<std::string>(1, request.tag() + "_timestamp"));
+
+  if (time_history.size() != 0) {
+    Timestamp latest_streamed_timestamp;
+    latest_streamed_timestamp.ParseFromString(time_history[0]);
+    if ((warble->timestamp().useconds() <=
+         latest_streamed_timestamp.useconds()) ||
+        (request.timestamp().useconds() > warble->timestamp().useconds())) {
+      delete warble;
```

Comment:
actually I think we should delete warble, but I am wondering if a user encounters high network delay when sometimes during the delay, another user sends a stream request, then some warbles are deleted in the database, then the first user loses some warbles, what do u think?

Reply:
You are right in that there is a chance that a single user might not receive all the warbles (since we delete the outdated ones). Here, there is a tradeoff between database space used and the reliability of the stream function (we could store all warbles with hashtag and have perfect reliability but high storage or vice-versa).

Since streaming is not a critical event (if I am on twitter and I see most of the hashtags I search for, I don't necessarily need to see all of them), I am okay with prioritizing storage. If it was a critical event like publish warble or registeruser, I would prioritize reliability.